File: src/Log4JParserCpp/Event.cpp
#include "Event.h"

namespace Log4JParser
{
    FixedString::FixedString (const char * value, const size_t size)
        : value_ (value), size_ (size)
    {

    }

    const char *FixedString::Value () const
    {
        return value_;
    }

    size_t FixedString::Size () const
    {
        return size_;
    }

    std::basic_ostream<char, std::char_traits<char>> &operator << (std::basic_ostream<char, std::char_traits<char>> &stream, FixedString &str)
    {
        stream.write (str.Value (), str.Size ());
        return stream;
    }

    Property::Property (const FixedString name, const FixedString value)
        : name_ (name), value_ (value)
    {

    }

    const FixedString Property::Name () const
    {
        return name_;
    }

    const FixedString Property::Value () const
    {
        return value_;
    }

    std::basic_ostream<char, std::char_traits<char>> &operator << (std::basic_ostream<char, std::char_traits<char>> &stream, Property &prop)
    {
        auto name = prop.Name ();
        auto value = prop.Value ();
        return stream << name << ": " << value;
    }

    Event::Event (Log4JEvent event)
        : event_ (event)
    {

    }

    Event::~Event ()
    {
        event_ = nullptr;
    }

    const FixedString Event::Level () const
    {
        const char *level;
        size_t levelSize;
        Log4JEventLevel (event_, &level, &levelSize);

        return FixedString (level, levelSize);
    }

    const FixedString Event::Logger () const
    {
        const char *logger;
        size_t loggerSize;
        Log4JEventLogger (event_, &logger, &loggerSize);

        return FixedString (logger, loggerSize);
    }

    const FixedString Event::Thread () const
    {
        const char *thread;
        size_t threadSize;
        Log4JEventThread (event_, &thread, &threadSize);

        return FixedString (thread, threadSize);
    }

    int64_t Event::Timestamp () const
    {
        return Log4JEventTimestamp (event_);
    }

    const FixedString Event::Message () const
    {
        const char *message;
        size_t messageSize;
        Log4JEventMessage (event_, &message, &messageSize);

        return FixedString (message, messageSize);
    }

    const FixedString Event::Throwable () const
    {
        const char *throwable;
        size_t throwableSize;
        Log4JEventThrowable (event_, &throwable, &throwableSize);

        return FixedString (throwable, throwableSize);
    }

    void Event::Properties (std::vector<Property> &properties) const
    {
        const size_t bufferSize = 16U;
        Log4JEventProperty buffer[bufferSize];

        size_t totalEvents;
        size_t totalEventsRead = 0U;
        do
        {
            totalEvents = Log4JEventProperties (event_, totalEventsRead, buffer, bufferSize);
            auto eventsRemaining = totalEvents - totalEventsRead;
            auto eventsRead = eventsRemaining > bufferSize ? bufferSize : eventsRemaining;

            for (size_t i = 0U; i < eventsRead; ++i)
            {
                FixedString name (buffer[i].name, buffer[i].nameSize);
                FixedString value (buffer[i].value, buffer[i].valueSize);
                properties.push_back (Property (name, value));
            }

            totalEventsRead += eventsRead;
        } while (totalEventsRead != totalEvents);
    }

    Log4JEvent Event::GetEvent (const Event event)
    {
        return event.event_;
    }

    EventSource::EventSource (char *xmlString)
    {
        Log4JEventSourceInitXmlString (&eventSource_, xmlString);
    }

    EventSource::~EventSource ()
    {
        Log4JEventSourceDestroy (eventSource_);
        eventSource_ = nullptr;
    }

    Event EventSource::First () const
    {
        auto event = Log4JEventSourceFirst (eventSource_);

        return Event (event);
    }

    Event EventSource::Next (Event event) const
    {
        auto cEvent = Log4JEventSourceNext (eventSource_, Event::GetEvent (event));
        return Event (cEvent);
    }

    const Log4JEventSource *EventSource::GetEventSource (const EventSource *eventSource)
    {
        return eventSource->eventSource_;
    }

    Log4JEventSource *EventSource::GetEventSource (EventSource *eventSource)
    {
        return eventSource->eventSource_;
    }
}

File: src/Log4JParserNet/Event.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;

namespace Log4JParserNet
{
    public sealed class Event
    {
        private readonly EventHandle impl_;

        private readonly ulong id_;

        private readonly IEventSource owner_;

        internal EventHandle Handle
        {
            get { return impl_; }
        }

        internal Event (EventHandle impl, ulong id, IEventSource owner)
        {
            Debug.Assert (impl != null, "Event.ctor: impl is null.");
            Debug.Assert (owner != null, "Event.ctor: owner is null.");

            impl_ = impl;
            id_ = id;
            owner_ = owner;
        }

        private unsafe string PtrToString (IntPtr value, UIntPtr size)
        {
            var intSize = (int) size.ToUInt32 ();
            string result = null;

            if (value != IntPtr.Zero)
            {
                var buffer = (byte*) value.ToPointer ();

                var encoding = owner_.Encoding;
                var charCount = encoding.GetCharCount (buffer, intSize);

                var decodeBuffer = new char[charCount];
                if (decodeBuffer.Length > 0)
                {
                    fixed (char* decodeBufferPtr = &decodeBuffer[0])
                    {
                        encoding.GetChars (buffer, intSize, decodeBufferPtr, decodeBuffer.Length);
                    }
                }

                result = new String (decodeBuffer);
            }

            return result;
        }

        public string Level
        {
            get
            {
                GuardState ();

                IntPtr value;
                UIntPtr size;
                Log4JParserC.Log4JEventLevel (impl_, out value, out size);

                return PtrToString (value, size);
            }
        }

        public string Logger
        {
            get
            {
                GuardState ();

                IntPtr value;
                UIntPtr size;
                Log4JParserC.Log4JEventLogger (impl_, out value, out size);

                return PtrToString (value, size);
            }
        }

        public string Thread
        {
            get
            {
                GuardState ();

                IntPtr value;
                UIntPtr size;
                Log4JParserC.Log4JEventThread (impl_, out value, out size);

                return PtrToString (value, size);
            }
        }

        public Int64 Timestamp
        {
            get
            {
                GuardState ();

                return Log4JParserC.Log4JEventTimestamp (impl_);
            }
        }

        public DateTime Time
        {
            get { return Log4JParserNet.Timestamp.ToDateTime (Timestamp); }
        }

        public string Message
        {
            get
            {
                GuardState ();

                IntPtr value;
                UIntPtr size;
                Log4JParserC.Log4JEventMessage (impl_, out value, out size);

                return PtrToString (value, size);
            }
        }

        public string Throwable
        {
            get
            {
                GuardState ();

                IntPtr value;
                UIntPtr size;
                Log4JParserC.Log4JEventThrowable (impl_, out value, out size);

                return PtrToString (value, size);
            }
        }

        public IList<KeyValuePair<string, string>> GetProperties ()
        {
            const int bufferSize = 16;
            var buffer = new Log4JEventProperty[bufferSize];

            var result = new List<KeyValuePair<string, string>> (16);

            uint totalEvents;
            var totalEventsRead = 0U;
            do
            {
                totalEvents = Log4JParserC
                    .Log4JEventProperties (impl_, new UIntPtr (totalEventsRead), buffer, new UIntPtr ((uint) buffer.Length))
                    .ToUInt32 ();

                var eventsRemaining = totalEvents - totalEventsRead;
                var eventsRead = eventsRemaining > bufferSize ? bufferSize : eventsRemaining;

                for (var i = 0U; i < eventsRead; ++i)
                {
                    var name = PtrToString (buffer[i].Name, buffer[i].NameSize);
                    var value = PtrToString (buffer[i].Value, buffer[i].ValueSize);
                    result.Add (new KeyValuePair<string, string> (name, value));
                }

                totalEventsRead += eventsRead;
            } while (totalEventsRead != totalEvents);

            return result;
        }

        public ulong Id
        {
            get { return id_; }
        }

        private void GuardState ()
        {
            if (owner_.IsInvalid)
            {
                throw new InvalidOperationException ("Owner object is invalid.");
            }
        }
    }
}
